lib/amqp_client.js

```diff
@@ -157,20 +159,47 @@ AMQPClient.prototype.connect = function(url, policyOverrides) {
     self._connection.on(Connection.Connected, function(c) {
       debug('connected');
       self.emit(AMQPClient.ConnectionOpened);
+
+      var promises = [];
+      
+      // Only map the session if it has already been created
       if (self._session) {
         debug('session already exists, re-using');
         self._session.connection = self._connection;
-      } else {
-        self._session = self._newSession(c);
+
+        promises.push(new Promise(function(resolve, reject) {
+          self._session.once(Session.Mapped, function(s) {
+            debug('client session mapped');
+            resolve();
+          });
+
+          self._session.begin(self._session.policy || self.policy.session);
```

Comment:
>self._session.policy || self.policy.session [](start = 30, length = 43)

if these 2 are the same, no problem. if the existing session doesn't have a policy object, apply the configured one. i get that part - but what if self._session.policy is set to something that is incompatible with self.policy.session (which i assume is the desirable policy) ? should these objects be merged instead ? or why keep the existing policy object at all?

Reply:
ha I see. the philosophy of policies as i understand it though is that they are merged and what is passed around are policy overrides - so starting with the client policy and then merging/overriding with the session policy would feel like the natural thing to do. but i defer to your judgement.

---
In reply to: [163131314](https://github.com/noodlefrenzy/node-amqp10/pull/332#discussion_r163131314) [](ancestors = 163131314)

Reply:
i think it's reasonable to do the merge on the built-in one (this line). the user one is weird because the `begin()` here is a hidden operation. the policy for the session is specified separately so i don't want to override it if it's already there. i'll add a comment to that one